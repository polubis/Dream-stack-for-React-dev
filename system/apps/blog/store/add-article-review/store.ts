import { create } from 'zustand';
import type { AddArticleReviewStore } from './defs';
import { addArticleReviewForm } from './core';

const useAddArticleReviewStore = create<AddArticleReviewStore.State>(() => ({
  is: 'idle',
  form: addArticleReviewForm.init({ content: '' }),
}));

export { useAddArticleReviewStore, addArticleReviewForm };
