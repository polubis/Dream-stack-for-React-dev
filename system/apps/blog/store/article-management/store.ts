import { create } from 'zustand';
import type * as ArticleManagement from './defs';

const useArticleManagementStore = create<ArticleManagement.State>(() => ({
  is: 'idle',
}));

export { useArticleManagementStore };
