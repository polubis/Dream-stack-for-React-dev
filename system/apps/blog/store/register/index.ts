export * from './register.store';
export * from './defs';
