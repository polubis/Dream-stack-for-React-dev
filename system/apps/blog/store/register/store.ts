import { create } from 'zustand';
import { RegisterStore } from './defs';
import { registerForm } from './core';

const useRegisterStore = create<RegisterStore.State>(() => ({
  is: 'idle',
  error: null,
  form: registerForm.init({
    login: '',
    email: '',
    password: '',
    confirmPassword: '',
  }),
}));

export { useRegisterStore };
