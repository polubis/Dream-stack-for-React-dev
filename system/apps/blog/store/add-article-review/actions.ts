import { createArticleReview, getError } from '@system/blog-api';
import type * as AddArticleReview from './defs';
import { addArticleReviewForm, useAddArticleReviewStore } from './store';

const { setState: set, getState: get } = useAddArticleReviewStore;

const add_article_review_actions: AddArticleReview.Actions = {
  setField: (key, value) => {
    set({ form: addArticleReviewForm.set(get().form)({ [key]: value }) });
  },
  confirm: async (id) => {
    const form = addArticleReviewForm.confirm(get().form);

    if (form.invalid) {
      set({ form });
      return;
    }

    set({ is: 'busy' });

    try {
      const response = await createArticleReview({
        id,
        ...get().form.values,
      });

      set({ is: 'ok' });
      add_article_review_actions.setField('content', '');

      return response;
    } catch (error: unknown) {
      set({ is: 'fail', error: getError(error) });
    }
  },
};

export { add_article_review_actions };
