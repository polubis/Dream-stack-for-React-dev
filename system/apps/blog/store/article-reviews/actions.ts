import { getArticleReviews, getError } from '@system/blog-api';
import type * as ArticleReviews from './defs';
import { useArticleReviewsStore } from './store';

const { setState: set, getState: get } = useArticleReviewsStore;

const article_reviews_actions: ArticleReviews.Actions = {
  load: async (id) => {
    set({ is: 'busy' }, true);

    try {
      const { data } = await getArticleReviews({ id });
      set({ is: 'ok', reviews: data }, true);
    } catch (error: unknown) {
      set({ is: 'fail', error: getError(error) }, true);
    }
  },
  addReview: (review) => {
    const state = get();

    if (state.is === 'ok') {
      set({ reviews: [...state.reviews, review] });
      return;
    }

    throw Error('You are trying to add review when there is no data yet');
  },
  reset: () => {
    set({ is: 'idle' }, true);
  },
};

export { article_reviews_actions };
