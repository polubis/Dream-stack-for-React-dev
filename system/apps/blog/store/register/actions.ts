import { getError, register } from '@system/blog-api';
import { type RegisterStore } from './defs';
import { useRegisterStore } from './store';
import { registerForm } from './core';

const { setState: set, getState: get } = useRegisterStore;

const register_store_actions: RegisterStore.Actions = {
  setField: (key, value) => {
    set({ form: registerForm.set(get().form)({ [key]: value }) });
  },
  submit: async () => {
    const form = registerForm.confirm(get().form);

    if (form.invalid) {
      set({ form });
      return;
    }

    set({ is: 'busy' });

    try {
      await register(useRegisterStore.getState().form.values);

      set({ is: 'ok' });
    } catch (error: unknown) {
      set({ is: 'fail', error: getError(error) });
    }
  },
};

export { register_store_actions };
