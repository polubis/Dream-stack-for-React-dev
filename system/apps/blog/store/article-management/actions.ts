import { useArticleManagementStore } from './store';
import { article_reviews_store_actions } from '../article-reviews';
import { article_store_actions } from '../article/actions';
import { add_article_review_store_actions } from '../add-article-review';
import { mockArticleReview } from '@system/blog-api-mocks';
import { change_article_status_store_actions } from '../change-article-status';
import type { ArticleManagementStore } from './defs';

const { setState: set } = useArticleManagementStore;

const article_management_store_actions: ArticleManagementStore.Actions = {
  start: (id, url, lang) => {
    set({ is: 'active', id });
    article_reviews_store_actions.load(id);
    article_store_actions.load({ url, lang });
  },
  sendForApproval: async (id) => {
    await change_article_status_store_actions.sendForApproval(id);
    article_store_actions.update({ status: 'WaitingForApproval' });
    return;
  },
  reset: () => {
    set({ is: 'idle' });
  },
  confirm: async (id) => {
    await add_article_review_store_actions.confirm(id);
    // @TODO: Remove mock when backend will finish work.
    article_reviews_store_actions.addReview(mockArticleReview());
  },
  changeStatus: async (id, status) => {
    if (status === 'Accepted') {
      await change_article_status_store_actions.accept(id);
      article_store_actions.update({ status });
      return;
    }

    if (status === 'NeedWork') {
      await change_article_status_store_actions.reject(id);
      article_store_actions.update({ status });
      return;
    }
  },
};

export { article_management_store_actions };
