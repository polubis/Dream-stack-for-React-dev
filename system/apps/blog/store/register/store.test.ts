import { act, waitFor } from '@testing-library/react';
import { useRegisterStore } from './store';
import { storeFixture } from '../test-utils';
import { register, getError } from '@system/blog-api';
import { mockRegisterPayload, mockResponseError } from '@system/blog-api-mocks';
import type { RegisterStore } from './defs';
import { register_store_actions } from './actions';

jest.mock('@system/blog-api');

describe('Register feature works when: ', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('initial state is correct', () => {
    const { result, restore } = storeFixture(useRegisterStore);

    expect(result.current.is).toBe('idle');
    expect(result.current).toMatchSnapshot();

    restore();
  });

  it('handles error', async () => {
    (register as jest.Mock).mockImplementation(() => Promise.reject());
    (getError as jest.Mock).mockImplementation(
      jest.requireActual('@system/blog-api')['getError']
    );

    const payload = mockRegisterPayload();
    const { result, restore } = storeFixture(useRegisterStore);

    expect(result.current.is).toBe('idle');

    act(() => {
      for (const key in payload) {
        register_store_actions.setField(
          key as keyof RegisterStore.FormData,
          payload[key]
        );
      }

      register_store_actions.submit();
    });

    expect(result.current.is).toBe('busy');

    await waitFor(() => {
      expect(register).toHaveBeenCalledTimes(1);
    });

    expect(result.current.is).toBe('fail');
    expect(result.current.error).toEqual(mockResponseError());

    restore();
  });

  it('allows to change field value', () => {
    const { result, restore } = storeFixture(useRegisterStore);

    expect(result.current.form.values.confirmPassword).toBe('');

    const value = 'd';

    act(() => {
      register_store_actions.setField('confirmPassword', value);
    });

    expect(result.current.form.values.confirmPassword).toBe(value);

    restore();
  });

  it('skips request call when form is invalid', () => {
    const { result, restore } = storeFixture(useRegisterStore);

    act(() => {
      register_store_actions.submit();
    });

    expect(result.current.form).toMatchSnapshot();

    restore();
  });

  it('allows to submit and validates', async () => {
    (register as jest.Mock).mockImplementation(() => Promise.resolve());

    const { result, restore } = storeFixture(useRegisterStore);
    const payload = mockRegisterPayload();

    act(() => {
      for (const key in payload) {
        register_store_actions.setField(
          key as keyof RegisterStore.FormData,
          payload[key]
        );
      }
    });

    for (const key in payload) {
      expect(result.current.form.values[key]).toBe(payload[key]);
    }

    act(() => {
      register_store_actions.submit();
    });

    expect(result.current.form).toMatchSnapshot();
    expect(result.current.is).toBe('busy');

    await waitFor(() => {
      expect(result.current.is).toBe('ok');
    });

    restore();
  });
});
