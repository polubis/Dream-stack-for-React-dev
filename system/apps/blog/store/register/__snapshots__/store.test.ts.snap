// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register feature works when:  allows to submit and validates 1`] = `
Object {
  "dirty": false,
  "errors": Object {
    "confirmPassword": null,
    "email": null,
    "login": null,
    "password": null,
  },
  "invalid": false,
  "invalidCount": 0,
  "keys": Array [
    "login",
    "email",
    "password",
    "confirmPassword",
  ],
  "pristine": true,
  "progress": 100,
  "result": Object {
    "confirmPassword": false,
    "email": false,
    "login": false,
    "password": false,
  },
  "touch": Object {
    "confirmPassword": true,
    "email": true,
    "login": true,
    "password": true,
  },
  "touched": true,
  "untouched": false,
  "valid": true,
  "validCount": 4,
  "values": Object {
    "confirmPassword": "piotr1994",
    "email": "piotr@wp.pl",
    "login": "piotr1994",
    "password": "piotr1994",
  },
}
`;

exports[`Register feature works when:  initial state is correct 1`] = `
Object {
  "error": null,
  "form": Object {
    "dirty": false,
    "errors": Object {
      "confirmPassword": null,
      "email": null,
      "login": null,
      "password": null,
    },
    "invalid": false,
    "invalidCount": 0,
    "keys": Array [
      "login",
      "email",
      "password",
      "confirmPassword",
    ],
    "pristine": true,
    "progress": 0,
    "result": Object {
      "confirmPassword": false,
      "email": false,
      "login": false,
      "password": false,
    },
    "touch": Object {
      "confirmPassword": false,
      "email": false,
      "login": false,
      "password": false,
    },
    "touched": false,
    "untouched": true,
    "valid": true,
    "validCount": 4,
    "values": Object {
      "confirmPassword": "",
      "email": "",
      "login": "",
      "password": "",
    },
  },
  "is": "idle",
}
`;

exports[`Register feature works when:  skips request call when form is invalid 1`] = `
Object {
  "dirty": true,
  "errors": Object {
    "confirmPassword": "required",
    "email": "required",
    "login": "required",
    "password": "required",
  },
  "invalid": true,
  "invalidCount": 4,
  "keys": Array [
    "login",
    "email",
    "password",
    "confirmPassword",
  ],
  "pristine": false,
  "progress": 0,
  "result": Object {
    "confirmPassword": true,
    "email": true,
    "login": true,
    "password": true,
  },
  "touch": Object {
    "confirmPassword": false,
    "email": false,
    "login": false,
    "password": false,
  },
  "touched": false,
  "untouched": true,
  "valid": false,
  "validCount": 0,
  "values": Object {
    "confirmPassword": "",
    "email": "",
    "login": "",
    "password": "",
  },
}
`;
